### ISSUE TITLE
How to handle older recipes?

### ISSUE THREAD

**Issue by daler** [bioconda/bioconda-recipes#6491]
As software is updated, we update recipes. What should happen to the older recipes? In some cases, we retain the older recipes in subdirectories; in others we overwrite the old with the new. Currently our policy at http://bioconda.github.io/guidelines.html#versions is to overwrite the old with the new but we haven't been enforcing this.

This recently came up in https://github.com/bioconda/bioconda-recipes/pull/5117 and https://github.com/bioconda/bioconda-recipes/pull/6471.

I think that overwriting old with new is OK since recipes are always bundled with the package tarball, and recipes will always be in the git repo history. We could build a tool that, given a package and the repo, will identify which commit the recipe came from. This also keeps the repo directory structure clean. However there is also value in retaining a more visible record of the recipe changes in the repo in the form of older recipe subdirectories.

Thoughts?

cc @yhoogstrate @karel-brinda @jerowe @bioconda/core

**Comment by jerowe** [comment]
I have a slight preference towards keeping the old recipe, but like you
said it's redundant information. I think I've only had one time I really
had to go looking for an old version of a recipe, and I was able to get
that from the git history.

So very slight personal preference, but I can also see why it's not
necessary or practical.

**Comment by yhoogstrate** [comment]
From a dev and git perspective it is more convenient to track all versions. Lets assume we have a tool which is superseded by a new version and the recipe is overwritten. Suddenly the recipe of the previous version needs a fix. Then the old versions recipe is taken from the archive, reinserted, and lost all its git history. I think file size is the least to worry about because they're quite small flat files. It has my preference to include recipes for all ( at least major) versions, ideally all versions. If we keep appropriate directory names en separate by versions, maybe even major nested with minor versions, it doesnt need to become chaotic.

**Comment by rvalieris** [comment]
imho, we should always update recipes in-place, keeping the old recipe is an exception for when we need to rebuild/update the older version *after* a newer version already exists, and this is not common occurrence.

thats not to say we shouldn't allow it, but I think its better to only keep older recipes we really need instead of all of them.

**Comment by karel-brinda** [comment]
I think the question is what's the final product of Bioconda – the binaries? the recipes? tarballs? the github repository (with its history)?

If "updating recipes in place" becomes the recommended policy, then I would be happy for an additional, automatically generated, Github repository with all the versions (of all the recipes for all  the packages) that were used to generate the binaries. Knowing the recipe for a given package is important for reproducibility.

Such a repository would probably have to be generated from the tarballs. Of course, everyone could do it by himself/herself if he/she needs it, but this would mean that many people would have to download all the tarballs (i.e., uncomfortableness + a lot of traffic).

**Comment by rvalieris** [comment]
keeping track of recipes for the sake of reproducibility is a nice idea, but in this case simply separating recipes by version won't be enough, we will need to store each version+build number, otherwise you will need to check the git history anyway.

@daler proposed we could have a tool that tracks which commit correspond to each package build, I think this is a better idea because it can be automatized to show up on bioconda.github.io for example.

**Comment by karel-brinda** [comment]
@rvalieris  Isn't usually sufficient to keep just the recipe with the highest build number?

**Comment by rvalieris** [comment]
I mean, it depends on the purpose, if we want reproducibility of the packages we need to keep the correspondent recipe of each package, and packages are identified by <name>-<version>_<build#>

**Comment by vinjana** [comment]
If you'd keep only newest versions, how would you deal with old software versions? I ported some of our pipelines to conda, which required me to add some old package versions to your repo (which was also accepted).

But if you keep only the newest build versions, would you keep the subdirectory for old versions with their newest build number, or would you delete them the next time someone updates the the most recent version, or would you even disallow adding builds for older software versions? ...

**Comment by bgruening** [comment]
Even if we only store the latest version in master, we need to have a mechanism to add older versions. So in this case we probably need to add subdirs. Should these then be removed afterwards? This seems wired to me. I kind of like the current model. Maintainers can decide this on there own. But if people want to keep old version, because they want to maintain them we keep them.

This would mean:
 * if there are subdirs keep them and don't remove them
 * if there are no subdirs, change in place
 * subdirs, if needed, should only be used for version-number, not build number (these can really be extracted from the tarballs or git history)

**Comment by daler** [comment]
I just want to point out that if you unpack a conda package's tarball, the recipe itself is always included in the `info/recipe` subdir, including the original yaml as well as the jinja2-rendered final version. It sounds like those in favor of enforcing subdirectories are (rightly!) concerned about knowing which recipe a package was generated from, but IMHO that's already handled by conda bundling the recipe in the package.

**Comment by rvalieris** [comment]
sorry, I completely forgot about this thread.
my only reservation here was on the idea of keeping older versions of all recipes, even when they aren't necessary.
I think adding subdirs increase the complexity of the repo, not for git, but for us humans, so we should try to not use it so often.

as an example, every time a new subdir is added on a pull request the review is harder because its not obvious what was changed, git thinks new whole files were added.

the idea of keeping track of recipes for the sake of reproducibility is nice but doing this tracking by hand is doomed to fail, we make mistakes and forget guidelines, but thats alright, thats why we should be making scripts that do the work for us.

I agree with your points @bgruening.

**Comment by johanneskoester** [comment]
I agree as well. The subdirs are for maintaining an older version in parallel to the current one. The tracking of the recipe used to build a version can be handled directly via the conda package as pointed out by @daler.

**Comment by bgruening** [comment]
I guess we settled on the subdir approach for now. Closing for now.
